multiraft/heartbeat_test.go

```diff
@@ -0,0 +1,152 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package multiraft
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+func processEventsUntil(ch <-chan *interceptMessage, f func(*RaftMessageRequest) bool) {
+	for e := range ch {
+		e.ack <- struct{}{}
+		if f(e.args.(*RaftMessageRequest)) {
+			return
+		}
+	}
+}
+
+// validateHeartbeatCount receives messages from the given channel and compares
+// the observed number of heartbeat requests and responses.
+// Once the channel closes or enough heartbeat responses have been observed,
+// the function returns.
+func validateHeartbeatCount(ch <-chan *interceptMessage,
+	heartbeatCount, heartbeatResponseCount int, t *testing.T) {
+	// maps nodeID to heartbeats sent by this node minus heartbeat
+	// responses received by this node.
+	outInDelta := make(map[uint64]int)
+	cntH := 0
+	cntR := 0
+	defer func() {
+		for nodeID, delta := range outInDelta {
+			// If one of the checks below fails, this information will be useful.
+			log.Infof("node %v: %d heartbeats without response", nodeID, delta)
+		}
+		if cntH != heartbeatCount {
+			t.Errorf("unexpected total number of heartbeats: %d (expected %d)",
+				cntH, heartbeatCount)
+		}
+		if cntR != heartbeatResponseCount {
+			t.Errorf("unexpected total number of heartbeat responses: %d (expected %d)",
+				cntR, heartbeatResponseCount)
+		}
+	}()
+	for iMsg := range ch {
+		iMsg.ack <- struct{}{}
+		req := iMsg.args.(*RaftMessageRequest)
+		switch req.Message.Type {
+		case raftpb.MsgHeartbeat:
+			cntH++
+			outInDelta[req.Message.From]++
+		case raftpb.MsgHeartbeatResp:
+			cntR++
+			nodeID := req.Message.To
+			outInDelta[nodeID]--
+			if outInDelta[nodeID] < 0 {
+				t.Errorf("node %v: more responses received than heartbeats sent",
+					nodeID)
+			}
+		}
+		// If we've already received all the heartbeat responses that
+		// we have asked for, return. Without this, it's more awkward
+		// to know when to shut down the cluster.
+		if cntR >= heartbeatResponseCount {
+			return
+		}
+	}
+}
+
+func TestHeartbeat(t *testing.T) {
+	nodeCount := 3
+	cluster := newTestCluster(nodeCount, t)
+	// Outgoing messages block the client until we acknowledge them, and
+	// messages are sent synchronously.
+	cluster.transport.EnableEvents()
+
+	cluster.start()
+	leaderTickCount := 18
+	// Since waitForElection() currently triggers two ticks, it also triggers
+	// some heartbeats which remain unanswered (the sender isn't leader yet).
+	unansweredHeartbeats := (nodeCount - 1) * 2 // 2 because of 2 election ticks.
```

Comment:
I could probably have an extra consumer goroutine that processes messages until it receives the second heartbeat, at which it will make room for the goroutine that's in the test now.
But that would mean adapting even more to the fact that currently we have to Tick() to get an election, and it would test less (the current setup tests that non-leader heartbeats actually get discarded, albeit it's somehow muddled up in the actual test)